GarNet/garnet.py

```diff
@@ -324,221 +242,54 @@ def TF_regression(motifs_and_genes_file_or_dataframe, expression_file, output_di
-		motifs_file (str or FILE): filepath or file object for the motifs file
-		options (dict): {"upstream_window": int, "downstream_window": int, "tss": bool, "output_dir": string (optional)})
-		kgXref_file (str or FILE): filepath or file object for the known genes reference file
-
-	Returns:
-		dataframe: A dataframe listing transcription factor binding motifs and nearby genes.
-	"""
-
-	reference = dict_of_IntervalTree_from_known_genes_file(known_genes_file, options)
-	motifs = dict_of_IntervalTree_from_motifs_file(motifs_file)
-
-	motifs_with_associated_genes = intersection_of_dict_of_intervaltree(motifs, reference)
-
-	motifs_and_genes = [{**motif, **gene} for motif, gene in motifs_with_associated_genes]
-
-	columns_to_output = ["chrom", "motifStart", "motifEnd", "motifName", "motifScore", "geneName", "geneStart", "geneEnd"]
-	motifs_and_genes = pd.DataFrame.from_records(motifs_and_genes, columns=columns_to_output)
-	motifs_and_genes['motif_to_gene_distance'] = motifs_and_genes['motifStart'] - motifs_and_genes['geneStart']
-
-	return motifs_and_genes
-
-
-def dict_of_IntervalTree_from_peak_file(peaks_file):
-	"""
-	Arguments:
-		peaks_file (str or FILE): filepath or FILE object
-
-	Returns:
-		dict: dictionary of intervals in known genes to intervals in peaks.
-	"""
-
-	logger.info('  - Parsing peaks file...')
-	peaks = parse_peaks_file(peaks_file)
-	peaks = group_by_chromosome(peaks)
-	logger.info('  - Parse complete, constructing IntervalTrees...')
-	peaks = {chrom: IntervalTree_from_peaks(chromosome_peaks) for chrom, chromosome_peaks in peaks.items()}
-
-	return peaks
+######################################## Contruct Garnet File ########################################
 
+def tss_from_bed(bed_file): 
 
-def dict_of_IntervalTree_from_known_genes_file(known_genes_file, options):
-	"""
-	Arguments:
-		known_genes_file (str or FILE): filepath or file object for the known_genes file
-		options (dict): options which may come from the argument parser.
+	output_file = bed_file.replace(".bed", ".tss.bed")
 
-	Returns:
-		dict: dictionary of chromosome to IntervalTree of known genes
-	"""
+	# taken from https://github.com/arq5x/bedtools-protocols/blob/master/bedtools.md
+	bed_to_tss_cmd = ''' cat %s | awk 'BEGIN{OFS=FS="\t"} \
+	       { if ($6 == "+") \
+	         { print $1,$2,$2+1,$4,$5,$6 } \
+	         else if ($6 == "-") \
+	         { print $1,$3-1,$3,$4,$5,$6 } \
+	       }' \
+	| sort -k1,1 -k2,2n \
+	| uniq \
+	> %s''' %(bed_file, output_file)
 
-	logger.info('  - Parsing known genes file...')
-	reference = parse_known_genes_file(known_genes_file, kgXref_file=options.get('kgXref_file'))
-	reference = group_by_chromosome(reference)
-	logger.info('  - Parse complete, constructing IntervalTrees...')
-	reference = {chrom: IntervalTree_from_reference(genes, options) for chrom, genes in reference.items()}
+	logger.info('  - Wrote TSS file to ' + output_file)
+	subprocess.call(bed_to_tss_cmd, shell=True)
```

Comment:
We should try to avoid subprocess calls. Is there no way to perform this command in pyBedTools?

Reply:
I looked for a while, couldn't find one.

Reply:
Sounds reasonable. We can merge this then look to replace it with python later.